File: flyweight/java/Flyweight.java
import java.util.HashMap;
import java.util.Map;

// 享元类
record ChessPieceUnit(int id, String text, Color color) {
    enum Color {
        RED, BLACK
    }
}

class ChessPieceUnitFactory {
    private static final Map<Integer, ChessPieceUnit> pieces = new HashMap<>();

    static {
        pieces.put(1, new ChessPieceUnit(1, "車", ChessPieceUnit.Color.BLACK));
        pieces.put(2, new ChessPieceUnit(2, "馬", ChessPieceUnit.Color.BLACK));
        pieces.put(3, new ChessPieceUnit(3, "象", ChessPieceUnit.Color.BLACK));
        pieces.put(4, new ChessPieceUnit(4, "士", ChessPieceUnit.Color.BLACK));
        pieces.put(5, new ChessPieceUnit(5, "將", ChessPieceUnit.Color.BLACK));
        pieces.put(6, new ChessPieceUnit(6, "士", ChessPieceUnit.Color.BLACK));
        pieces.put(7, new ChessPieceUnit(7, "象", ChessPieceUnit.Color.BLACK));
        pieces.put(8, new ChessPieceUnit(8, "馬", ChessPieceUnit.Color.BLACK));
        pieces.put(9, new ChessPieceUnit(9, "車", ChessPieceUnit.Color.BLACK));
        pieces.put(10, new ChessPieceUnit(10, "砲", ChessPieceUnit.Color.BLACK));
        pieces.put(11, new ChessPieceUnit(11, "砲", ChessPieceUnit.Color.BLACK));
        pieces.put(12, new ChessPieceUnit(12, "卒", ChessPieceUnit.Color.BLACK));
        pieces.put(13, new ChessPieceUnit(13, "卒", ChessPieceUnit.Color.BLACK));
        pieces.put(14, new ChessPieceUnit(14, "卒", ChessPieceUnit.Color.BLACK));
        pieces.put(15, new ChessPieceUnit(15, "卒", ChessPieceUnit.Color.BLACK));
        pieces.put(16, new ChessPieceUnit(16, "卒", ChessPieceUnit.Color.BLACK));
        pieces.put(17, new ChessPieceUnit(17, "俥", ChessPieceUnit.Color.RED));
        pieces.put(18, new ChessPieceUnit(18, "傌", ChessPieceUnit.Color.RED));
        pieces.put(19, new ChessPieceUnit(19, "相", ChessPieceUnit.Color.RED));
        pieces.put(20, new ChessPieceUnit(20, "仕", ChessPieceUnit.Color.RED));
        pieces.put(21, new ChessPieceUnit(21, "帥", ChessPieceUnit.Color.RED));
        pieces.put(22, new ChessPieceUnit(22, "仕", ChessPieceUnit.Color.RED));
        pieces.put(23, new ChessPieceUnit(23, "相", ChessPieceUnit.Color.RED));
        pieces.put(24, new ChessPieceUnit(24, "傌", ChessPieceUnit.Color.RED));
        pieces.put(25, new ChessPieceUnit(25, "俥", ChessPieceUnit.Color.RED));
        pieces.put(26, new ChessPieceUnit(26, "炮", ChessPieceUnit.Color.RED));
        pieces.put(27, new ChessPieceUnit(27, "炮", ChessPieceUnit.Color.RED));
        pieces.put(28, new ChessPieceUnit(28, "兵", ChessPieceUnit.Color.RED));
        pieces.put(29, new ChessPieceUnit(29, "兵", ChessPieceUnit.Color.RED));
        pieces.put(30, new ChessPieceUnit(30, "兵", ChessPieceUnit.Color.RED));
        pieces.put(31, new ChessPieceUnit(31, "兵", ChessPieceUnit.Color.RED));
        pieces.put(32, new ChessPieceUnit(32, "兵", ChessPieceUnit.Color.RED));
    }

    public static ChessPieceUnit getChessPiece(int chessPieceId) {
        return pieces.get(chessPieceId);
    }
}

// ============================================================

class ChessPiece {
    private ChessPieceUnit chessPieceUnit;
    private int positionX;
    private int positionY;

    public ChessPiece(ChessPieceUnit chessPieceUnit, int positionX, int positionY) {
        this.chessPieceUnit = chessPieceUnit;
        this.positionX = positionX;
        this.positionY = positionY;
    }

    public ChessPieceUnit getChessPieceUnit() {
        return chessPieceUnit;
    }

    public void setChessPieceUnit(ChessPieceUnit chessPieceUnit) {
        this.chessPieceUnit = chessPieceUnit;
    }

    public int getPositionX() {
        return positionX;
    }

    public void setPositionX(int positionX) {
        this.positionX = positionX;
    }

    public int getPositionY() {
        return positionY;
    }

    public void setPositionY(int positionY) {
        this.positionY = positionY;
    }
}

class ChessBoard {
    private final Map<Integer, ChessPiece> chessPieces = new HashMap<>();

    public ChessBoard() {
        init();
    }

    private void init() {
        chessPieces.put(1, new ChessPiece(ChessPieceUnitFactory.getChessPiece(1), 0, 0));
        chessPieces.put(2, new ChessPiece(ChessPieceUnitFactory.getChessPiece(2), 1, 0));
        // ...省略摆放其他棋子的代码...
    }

    // Only for tests!
    public Map<Integer, ChessPiece> getChessPieces() {
        return chessPieces;
    }

    public void move(int chessPieceId, int toPositionX, int toPositionY) {
        ChessPiece chessPiece = chessPieces.get(chessPieceId);
        if (chessPiece != null) {
            chessPiece.setPositionX(toPositionX);
            chessPiece.setPositionY(toPositionY);
        }
    }
}

File: flyweight/go/flyweight.go
package flyweight

type Color int

const (
	RED Color = iota
	BLACK
)

type ChessPieceUnit struct {
	ID    int
	Text  string
	Color Color
}

var pieces = map[int]*ChessPieceUnit{
	1:  {1, "車", BLACK},
	2:  {2, "馬", BLACK},
	3:  {3, "象", BLACK},
	4:  {4, "士", BLACK},
	5:  {5, "將", BLACK},
	6:  {6, "士", BLACK},
	7:  {7, "象", BLACK},
	8:  {8, "馬", BLACK},
	9:  {9, "車", BLACK},
	10: {10, "砲", BLACK},
	11: {11, "砲", BLACK},
	12: {12, "卒", BLACK},
	13: {13, "卒", BLACK},
	14: {14, "卒", BLACK},
	15: {15, "卒", BLACK},
	16: {16, "卒", BLACK},
	17: {17, "俥", RED},
	18: {18, "傌", RED},
	19: {19, "相", RED},
	20: {20, "仕", RED},
	21: {21, "帥", RED},
	22: {22, "仕", RED},
	23: {23, "相", RED},
	24: {24, "傌", RED},
	25: {25, "俥", RED},
	26: {26, "炮", RED},
	27: {27, "炮", RED},
	28: {28, "兵", RED},
	29: {29, "兵", RED},
	30: {30, "兵", RED},
	31: {31, "兵", RED},
	32: {32, "兵", RED},
}

func getChessPiece(chessPieceID int) *ChessPieceUnit {
	return pieces[chessPieceID]
}

// ============================================================

type ChessPiece struct {
	*ChessPieceUnit
	PositionX int
	PositionY int
}

type ChessBoard struct {
	chessPieces map[int]*ChessPiece
}

func NewChessBoard() *ChessBoard {
	return &ChessBoard{
		chessPieces: map[int]*ChessPiece{
			1: {getChessPiece(1), 0, 0},
			2: {getChessPiece(2), 1, 0},
			// ...省略摆放其他棋子的代码...
		},
	}
}

// GetChessPieces Only for tests!
func (cb ChessBoard) GetChessPieces() map[int]*ChessPiece {
	return cb.chessPieces
}

func (cb ChessBoard) Move(chessPieceID, toPositionX, toPositionY int) {
	cb.chessPieces[chessPieceID].PositionX = toPositionX
	cb.chessPieces[chessPieceID].PositionY = toPositionY
}
